packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts

```diff
@@ -0,0 +1,267 @@
+                        args: ["--vimgrep", "-S", "--color=always", "-e", what, root],
+                };
+                const process: RawProcess = this.rawProcessFactory(processOptions);
+                this.ongoingSearches.set(searchId, process);
+
+                process.onError(error => {
+                        // tslint:disable-next-line:no-any
+                        let errorCode = (error as any).code;
+
+                        /* Try to provide somewhat clearer error messages, if possible.  */
+                        if (errorCode === 'ENOENT') {
+                                errorCode = 'could not find the ripgrep (rg) binary';
+                        } else if (errorCode === 'EACCES') {
+                                errorCode = 'could not execute the ripgrep (rg) binary';
+                        }
+
+                        const errorStr = `An error happened while searching (${errorCode}).`;
+                        this.wrapUpSearch(searchId, errorStr);
+                });
+
+                /* Running counter of results.  */
+                let numResults = 0;
+
+                /* Buffer to accumulate incoming output.  */
+                let databuf: string = "";
+
+                const lastMatch = {
+                        file: '',
+                        line: 0,
+                        index: 0,
+                };
+
+                process.output.on('data', (chunk: string) => {
+                        /* We might have already reached the max number of
+                           results, sent a TERM signal to rg, but we still get
+                           the data that was already output in the mean time.
+                           It's not necessary to return early here (the check
+                           for maxResults below would avoid sending extra
+                           results), but it avoids doing unnecessary work.  */
+                        if (opts && opts.maxResults && numResults >= opts.maxResults) {
+                                return;
+                        }
+
+                        databuf += chunk;
+
+                        while (1) {
+                                /* Check if we have a complete line.  */
+                                const eolIdx = databuf.indexOf('\n');
+                                if (eolIdx < 0) {
+                                        break;
+                                }
+
+                                /* Get and remove the line from the data buffer.  */
+                                let lineBuf = databuf.slice(0, eolIdx);
+                                databuf = databuf.slice(eolIdx + 1);
+
+                                /* Extract the various fields using the ANSI
+                                   control characters for colors as guides.  */
+
+                                /* Extract filename (magenta).  */
+                                const filenameRE = /^\x1b\[m\x1b\[35m(.+?)\x1b\[m:/;
```

Comment:
would be nice to extract the color codes from the string so we can change it easily.

Reply:
I did this:

```
        private readonly LINE_START = '^\x1b\\[m\x1b\\[32m';
        private readonly LINE_END = '\x1b\\[m:';
        ...
        const lineRE = new RegExp(`${this.LINE_START}(\\d+)${this.LINE_END}`);

```
I think that we the necessary escaping, it makes it a bit more error-prone, but I don't really mind.

Reply:
I think you could remove some escaping by using ```` ?

Reply:
I don't think so, `` strings still process backslashes.  So a `\[` would still become a `[` and be wrongly interpreted as the opening of a regex character set.